client/change.go

```diff
@@ -65,6 +65,10 @@ type Task struct {
 
 	SpawnTime time.Time `json:"spawn-time,omitempty"`
 	ReadyTime time.Time `json:"ready-time,omitempty"`
+
+	SnapName     string `json:"snap-name,omitempty"`
+	InstanceName string `json:"instance-name,omitempty"`
+	Revision     string `json:"revision,omitempty"`
```

Comment:
we don't want to add this kind of data this way. We probably want to add support instead for something like api-data that we have for Changes.  Also (even) if we do that the closest thing we can do that doesn't feel too forced would be to expose a list of instance names "affected-snaps" that matches the same concept in snapstate conflict logic, we would have to expose some of that to daemon

Reply:
Separately, how are the consumers expected to use this information? Say if a change touches more than one, would they list all the tasks, and then continue observing which ones for which snap are done?

Reply:
@bboozzoo consumer is expected to poll the associated refresh change, if it happens to include multiple snaps then filtering on the task level is needed.

Check the "Proposed snap-run during refresh UX flow" section in [SD167](https://docs.google.com/document/d/1HJQWKzgaB3yPKwRUqlxYhgyaKG5IcJe8eZOT4YY1CI8/edit#heading=h.v228s8hnqd53).

Reply:
You can see a video with what I have right now: https://warthogs.atlassian.net/browse/UDENG-2282 , in the most recent comment. As you can see, the idea is to have a progress bar for each snap being updated, and that requires to know how many tasks are assigned to that snap refresh, and how many are already done. Currently I'm presuming that there is one change per snap, so each change has only one snap in its _snap_names_ property, and thus all the tasks belong to that snap. But if there are several snaps in the same change, then each progress bar will show the same progress, unless this patch (or something similar) is implemented, that allows me to separate the tasks.

Reply:
Also, here is an screen capture
![imagen](https://github.com/snapcore/snapd/assets/105285003/52735a1b-f1e5-4847-b4e4-87d6b6bd7cd7)

Reply:
If you watch the video, you will notice that the notifications are shown twice. That's because I'm running snapd-desktop-integration from outside a snap, so snapd doesn't know that there is a monitoring program and still sends its own notifications.

Reply:
Of course, I want now to integrate all the progress bars in a single window.

Reply:
@pedronis The point is that each _change notify_ just contains a _change ID_, so I have to ask for that change itself using _GET CHANGE?id=XXX_. That change has a list of snaps affected by that change, and also a list of tasks associated with that change, and there is where I want to add this extra data.

Reply:
that really doesn't affect my proposal, All client.Task should have the data, so that seems fine

Reply:
Oh, so you mean to add a "data" field to the task, and add there that info?

Reply:
Anyway, as I said in the top, if it is guaranteed that, when doing a refresh after the user closes the application, only one snap is associated with each change, this won't be needed. I tried killing two applications "at the same time" using kill XXXXX && kill YYYYY, and still receive two changes, one per snap...

Reply:
maybe we could extend the change level `"data"` field to avoid duplicating the same information over and over again in each task? we currently attach this to requests to /v2/change/<id> (I believe those are affected snap names):

```
    "data": {
      "snap-names": [
        "core18",
        "core20",
        "core22",
        "core24",
        "gnome-42-2204",
        "google-cloud-sdk",
        "ifacetool",
        "lxd",
        "node",
        "ohmygiraffe",
        "snapcraft",
        "snapd",
        "the-force-engine",
        "toolbox",
        "ubuntu-image"
      ]
    }
```
what if we also add something like:

```
"snap-tasks": { 
    "<name>", [ <task IDs> ]
   // eg.
   "lxd": ["12", "13", "14"
 }
```

not sure if this would make the client side handling easier or more complicated, `tasks` is already a flat list of objects which makes it a bit inconvenient

Reply:
It looks like a good idea.

Reply:
Objections...?

Reply:
unclear that "snap-tasks" is a good name for it, so we don't have a good name, also it's not an obvious approach except is compact

Reply:
also the obvious way to attach the information in snap-tasks form would be when we set snap-names, but actually a Change can grow more tasks after that point

Reply:
> but actually a Change can grow more tasks after that point

to reiterate on this, snap can pull prerequisites (like a new base) or a re-refresh may occur, at which point the task list will grow again, so any progress bar the user sees will need to be adjusted such that it doesn't jump backwards

Reply:
> Anyway, as I said in the top, if it is guaranteed that, when doing a refresh after the user closes the application, only one snap is associated with each change, this won't be needed. I tried killing two applications "at the same time" using kill XXXXX && kill YYYYY, and still receive two changes, one per snap...

There is still another scenario where `snap run` of a certain snap is inhibited from running due to a normal refresh/auto-refresh that includes that snap and the progress UI should only be relevant to the snap that is being started and not the other snaps in the auto-refresh. It not just when running the snap after closing.